### PR TITLE
Logistic skat

### PR DESCRIPTION
**Pull request by charliethebrit**
Includes the logistic case of the SKAT algorithm. Changes include updates to the SKAT, SKATSuite, and SKATmodel Scala files including updating python tests and other documentation. 

Another significant change in the code is that the tests in SKATSuite have been re-written to ensure less code is duplicated while also added in permutation and noise tests for the skat algorithm which are not apart of the CI testing.

### CODE REVIEW

**Comment by jbloom22** [comment]
I've merged a stack of changes to this branch including:
- Cleaned up tests, including refactoring, making Balding-Nichols deterministic and removing lots of extra test code
- Reorder args in Scala to match Python, related bug fixes
- Improved large N performance by using single array D rather than A and B
- Moved dense versus sparse matching outside of loop
- Improved Python docs and Scala remarks
- Debugged test failure only in Spark 2.1.0, which turned out to be related to accuracy of Davies. I've increased accuracy to 1e-8 which is enough to make current tests pass.

Once this goes in, I'll make PRs to:
- Allow users to set accuracy and iterations on Davies, will use same defaults as R: 1e-6 and 10k.
- Add number of variants per key as column.
- Fix behavior to finish running even if some groups are too big upper bound, or if Cholesky fails.  Document this behavior.

Less urgently, but to keep in mind:
- If bottleneck, improve performance of Gramian computation in large N case using blocking
- Improve Davies C code

**Comment by jbloom22** [comment]
Dan is adding a symmetrize function to RichDenseMatrixDouble which I'll invoke to remove the symmetrizing hack on Cholesky.

File: python/hail/dataset.py [hunk 1]
```diff
@@ -4966,12 +4966,13 @@ def delete_va_attribute(self, ann_path, attribute):
     @record_method
     @typecheck_method(variant_keys=strlike,
                       single_key=bool,
-                      weight_expr=nullable(strlike),
                       y=strlike,
                       covariates=listof(strlike),
+                      weight_expr=nullable(strlike),
+                      use_logistic=bool,
```
Comment (tpoterba) [inline python/hail/dataset.py]: maybe just `logistic` as the param name?

File: python/hail/dataset.py [hunk 2]
```diff
@@ -5042,9 +5051,8 @@ def skat(self, variant_keys, single_key, y, weight_expr=None, covariates=[], use
 
         .. caution::
 
-            If a p-value of 0.0 is returned, then the true p-value falls in the range between 0.0 and
-            :math:`2^{-52} \\approx  2.22 \\cdot 10^{-16}` (machine epsilon).  This is because Davies method calculates the
-            left-hand area, which we then subtract from 1 to get the p-value.
+          Davies algorithm iterates up to 100k times until an accuracy of 1e-8 is achieved. Hence a reported p-value of
```
Comment (tpoterba) [inline python/hail/dataset.py]: The Davies algorithm

File: python/hail/dataset.py [hunk 3]
```diff
@@ -5053,22 +5061,23 @@ def skat(self, variant_keys, single_key, y, weight_expr=None, covariates=[], use
...
-        :param str weight_expr: Variant expression of numeric type for SKAT weight. When no weight is provided, weights are
-                           generated to from a beta distribution (alpha = 1, beta = 25) evalulated at the minor allele
-                           frequency of each variant.
+        :param weight_expr: Variant expression of numeric type for SKAT weights.
+        :type weight_expr: str or None
 
         :param covariates: List of covariate expressions.
         :type covariates: List of str
 
+        :param bool use_logistic: If true, runs logistic SKAT rather than linear SKAT.
```
Comment (tpoterba) [inline python/hail/dataset.py]: Use logistic test rather than linear test.

File: src/main/scala/is/hail/methods/Skat.scala [hunk 4]
```diff
@@ -4,40 +4,46 @@ import is.hail.utils._
...
     yExpr: String,
     covExpr: Array[String],
+    weightExpr: Option[String],
+    useLogistic: Boolean = false,
     useDosages: Boolean,
     useLargeN: Boolean = false): KeyTable = {
     val (y, cov, completeSampleIndex) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+
+    if (useLogistic && (!y.forall(yi => yi == 0d || yi == 1d)))
+      fatal(s"For logistic SKAT, phenotype must be Boolean or numeric with all present values equal to 0 or 1")
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Skat.scala]: shouldn't this write the offending phenotype?

File: src/main/scala/is/hail/methods/Skat.scala [hunk 5]
```diff
@@ -4,40 +4,46 @@ import is.hail.utils._
...
+
     val n = y.size
     val sampleMask = Array.fill[Boolean](vds.nSamples)(false)
     completeSampleIndex.foreach(i => sampleMask(i) = true)
     val filteredVds = vds.filterSamplesMask(sampleMask)
 
-    def computeSkat(keyedRdd: RDD[(Annotation, Iterable[(Vector[Double], Double)])], keyType: Type,
+    def computeSkatLinear(keyedRdd: RDD[(Any, Iterable[(Vector[Double], Double)])], keyType: Type,
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Skat.scala]: why change this back from annotation?

File: src/main/scala/is/hail/methods/Skat.scala [hunk 6]
```diff
@@ -58,21 +64,73 @@ object Skat {
...
+      XVX.forceSymmetry()
+      val Cinv = inv(cholesky(XVX))
+      val res = y - mu
+
+      val sc = keyedRdd.sparkContext
+      val sqrtVBc = sc.broadcast(sqrt(V))
+      val resBc = sc.broadcast(res)
+      val CinvXtVBc = sc.broadcast(Cinv * VX.t)
+
+      def variantPreProcess(gs: Vector[Double], w: Double): SkatTuple = {
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Skat.scala]: copy the `preprocessGenotypes` method name from above

**Comment by jbloom22** [comment]
back to you

**Pull request merged**